File: solutions/_34_find_first_and_last_position_of_element_in_sorted_array/solution.py
from typing import List


class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        start = self.search(nums, target, True)
        end = self.search(nums, target, False)
        return [start, end]

    # leftBias=[True/False], if false, res is rightBiased
    def search(self, nums, target, leftBias):
        L = 0
        R = len(nums) - 1
        i = -1
        while L <= R:
            mid = (L + R) // 2
            if target > nums[mid]:
                L = mid + 1
            elif target < nums[mid]:
                R = mid - 1
            else:
                i = mid
                if leftBias:
                    R = mid - 1
                else:
                    L = mid + 1
        return i


# https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/

File: solutions/_34_find_first_and_last_position_of_element_in_sorted_array/solution.java

package _34_find_first_and_last_position_of_element_in_sorted_array;

class Solution {
    public int[] searchRange(int[] nums, int target) {
        int start = search(nums, target, true);
        int end = search(nums, target, false);
        return new int[] { start, end };
    }

    private static int search(int[] nums, int target, boolean leftBias) {
        int L = 0;
        int R = nums.length - 1;
        int i = -1;
        while (L <= R) {
            int mid = L + (R - L) / 2;
            if (target > nums[mid]) {
                L = mid + 1;
            } else if (target < nums[mid]) {
                R = mid - 1;
            } else {
                i = mid;
                if (leftBias) {
                    R = mid - 1;
                } else {
                    L = mid + 1;
                }
            }
        }
        return i;
    }
}

/*
 * https://leetcode.com/problems/find-first-and-last-position-of-element-in-
 * sorted-array/
 */